### ISSUE TITLE
JP2000 returning OPJ_CLRSPC_UNKNOWN color space

### ISSUE THREAD

**Issue by Jehan** [uclouvain/openjpeg#1103]
Hi,

I came accross this image which is even in the external references on Wikipedia for the JPEG2000 format page: http://www.fnordware.com/j2k/jp2samples.html

When loading this file with OpenJPEG, the color space is OPJ_CLRSPC_UNKNOWN.
Is the JP2 file wrongly encoded or is there some feature missing in OpenJPEG?

For the background, we are currently porting our JPEG2000 support in GIMP to using OpenJPEG. This is where I came into this file when testing various reference images found on the web.

**Comment by boxerab** [comment]
This image has an embedded ICC profile, so the colour space info is stored in the profile and is not specified elsewhere in the JPEG 2000 header.  However, the ICC profile is corrupt, so it cannot be used. If you open this image with Kakadu kdu_show, the colours are messed up, due to corrupt ICC.

**Comment by Jehan** [comment]
Ok. So basically, we should not try to support it. If GIMP cannot open this file, that is correct behavior, isn't it?

**Comment by boxerab** [comment]
Yes, it is normal that OPJ_CLRSPC_UNKNOWN is returned. Not sure about the answers to your other questions.

**Comment by Jehan** [comment]
> Not sure about the answers to your other questions.

Well mostly wondering how a program should handle or display such a file.
Is there an acceptable fallback or is it just broken beyond repair? :-)

**Comment by szukw000** [comment]
@Jehan , @boxerab ,
gimp-2.8.22 shows the file relax.jp2 . IrfanView shows the file relax.jp2 .
animate shows the file relax.jp2 . Only kdu_show fails to show relax.jp2 .
winfried

**Comment by Jehan** [comment]
>  gimp-2.8.22 shows the file relax.jp2.

For info, this was using Jasper, back in 2.8.x. GIMP 2.10 will stop being able to show this file as we are moving to OpenJPEG, hence the bug report.

**Comment by boxerab** [comment]
Only kdu_show tries to apply the ICC profile, so it is the only program that has problems with this particular file. For Gimp, you will need to find a way of retrieving ICC from the compressed JPEG 2000 file. I don't think OpenJPEG currently does this.

**Comment by Jehan** [comment]
Ok. So what we do now in GIMP (current master) is that we determine the color space from the embedded ICC profile if it is not declared in the header. So now we can cleanly load this image, though the colors are indeed all fucked up, which is apparently normal since you said that the ICC profile is broken. We can get the proper image assuming some sRGB by just removing the profile.
Unless you think we are doing something wrong (hopefully not, since I followed what I understood from your indications), this issue can be considered solved, I guess.

I still have some questions regarding color spaces that I encountered for other images. Rather than opening other issues, I guess I can just ask here. :-)

(1) So I was opening your various openjpeg-data conformance images. I noticed that images in JPEG 2000 codestream in particular (.j2c) all have no color space declared, neither have they any ICC profile embedded. Reading more, I am guessing this may be normal. Do codestream files even have a header? From what I could gather, the codestream format may just be the compressed image data, whereas the JP2 format is more of a container which includes JP2 codestream.
So can't JP2 codestream files have any color space or profile information?

Right now, our code is stupid and will basically assume some Grayscale or RGB/RGBA, depending on the number of components, when there is no color space declared, nor profile, in such a file, but obviously this is not right, since JPEG2000 supports much more color spaces (CMYK, YUV, YCC…). So should we just pop-up a dialog asking for the user to select the right format? Or is there another way in the format to detect the right color space?

(2) Apart from the ugly "assuming RGB/Gray" spaces, GIMP is now able to load most of the .j2c files in `conformance/` but for `a6_mono_colr.j2c`. We encountered here a case where some components did not have the same dimension (width, height) as the others. I am trying to understand what this means. Is this a RGB file? And if so, how can a component have different size as the others?
Or is it another kind of features of JPEG2000? If so, which?

I tried to look in `tests/conformance/CMakeLists.txt` and see that you are doing specific tests for this image, but I'm still unsure of its contents. It doesn't look like you have any document explaining what kind of data is in each file; this would be quite a useful document. :-)

**Comment by boxerab** [comment]
@Jehan  that sounds like an improvement if you are now looking at the ICC profile. But, you should be able to do a sanity check on the profile and determine that it is corrupt.  Check out Little CMS - excellent little library, if you are able to add a dependency.

Yes, `.j2k` only stores the compressed code stream, so no colour information. A dialog would be appropriate here.  You would need to handle alpha channels as well; also it is possible for the channels to be permuted i.e. the first channel of an RGB image may not be R. But this situation is probably pretty rare.

For `a6_mono_colr.j2c`, the channels are sub-sampled.   
https://en.wikipedia.org/wiki/Chroma_subsampling

**Comment by Jehan** [comment]
> that sounds like an improvement if you are now looking at the ICC profile

We were actually already extracting and applying the profile (GIMP is deeply color-managed nowadays, so applying the embedded profiles are something we do for every formats). We were simply not using it to determine the color space in JP2 since we were relying on the header data. The fact that sometimes the header data is not set is a bit weird. Anyway…

> Check out Little CMS - excellent little library, if you are able to add a dependency.

We already have a mandatory dependency to lcms2 (as I said, we are deeply color-managed in every piece of code). I'll have a look if there is anything in lcms2 API to detect the corruption but we actually already create a lcms profile with this data, and we get zero complaint from lcms with the profile embedded in this image.
I'll have a closer look at lcms2 API to see if there are other functions to detect the corruption, but so far I could not find any.

> For a6_mono_colr.j2c, the channels are sub-sampled.

Thanks. We'll try to have a closer look at this.

**Comment by boxerab** [comment]
Cool. When I run the  (C++11) code below  on the ICC profile for that image, I get bufferSize equal to zero.  I take this as a sign that the profile is corrupt.


	if (image->icc_profile_buf && image->icc_profile_len)	{
		auto in_prof = cmsOpenProfileFromMem(image->icc_profile_buf, image->icc_profile_len);
		if (in_prof) {
			cmsUInt32Number bufferSize = cmsGetProfileInfo(in_prof, cmsInfoDescription, cmsNoLanguage, cmsNoCountry, nullptr, 0);
			if (bufferSize) {
                                 //read profile name
                         }
                }
       }

**Comment by boxerab** [comment]
So, it looks like the description tag is missing.  I am not sure if this is a mandatory tag for a well-formed profile.

**Comment by Jehan** [comment]
> So, it looks like the description tag is missing. I am not sure if this is a mandatory tag for a well-formed profile. If not mandatory, then this does not actually indicate profile corruption. You may then have to apply the profile.

Yeah I was wondering the same thing, and looking through lcms code to determine if that were really mandatory. Maybe I'll open a bug report at lcms and ask them if this profile is really considered corrupted for them, and if so, how do we determine this (and also maybe cmsOpenProfileFromMem() should return NULL directly in that case).

**Comment by boxerab** [comment]
By the way, when trying to decompress this file to png, libpng complains that this ICC profile has an invalid rendering intent (intent equals 0xFFFF ).  So, this may be a clue that the ICC profile is invalid.

**Comment by Jehan** [comment]
So I encountered cases where the JP2 header had no color space, but the color space information was there through an ICC profile. And apparently J2K files never have neither header nor ICC profile.

Can there also be a JP2 with no color space defined in the header nor any ICC profile as well? Or are JP2 always expected to have one or the other?

**Comment by boxerab** [comment]
JP2 will have a colour space defined.

**Comment by Jehan** [comment]
So I am trying to figure out the best user-visible names for color spaces in the dialog when opening a JP2 codestream but I am a bit unsure.
I was relying on the comment part of the enum:

>    OPJ_CLRSPC_SYCC = 3,        /**< YUV */
>    OPJ_CLRSPC_EYCC = 4,        /**< e-YCC */

So I was calling them YUV and e-YCC.

First for YUV, reading Wikipedia I wonder if Y'UV is not actually the most common naming for the color space. Also I can't help but notice that the enum is *_SYCC. Searching Wikipedia again, I find this info:

> Amendment A1:2003 to IEC 61966-2-1:1999 describes an analogous sYCC encoding for YCbCr color spaces, an extended-gamut RGB encoding, and a CIELAB transformation.

So should it actually be sYCC?

As for e-YCC, I just cannot find any reference of it anywhere (under e-YCC or eYCC or anything). Is there maybe a more common name for this color space?

Sorry for all the (probably) stupid questions, but this is so hard to find information for these by not being freely published standards (actually more searches I do for these 2 color spaces basically return me links to OpenJPEG!).

**Comment by boxerab** [comment]
I think s refers to standard YCC, and e refers to extended YCC.

**Comment by boxerab** [comment]
np, looking forward to trying out the new version when it gets unleashed.

**Comment by jubalh** [comment]
Sorry for commenting on such an old issue. Just have a question about it.

For porting some application from jasper to digikam I read GIMPs code, as an example.

@Jehan how comes that GIMP displays http://www.fnordware.com/j2k/jp2samples.html for a split second correctly and then displays it weirdly (green)? I understood it like the corrupted color profile makes you not really support it but don't get how it can display it 'correctly' at first.

**Comment by Jehan** [comment]
@jubalh 
> @Jehan how comes that GIMP displays http://www.fnordware.com/j2k/jp2samples.html for a split second correctly and then displays it weirdly (green)?

Probably you see it fine for this split second before the profile is applied.
As I explained in an [early comment](https://github.com/uclouvain/openjpeg/issues/1103#issuecomment-371987364), the way to see an acceptable image is by loading the image while keeping the profile (don't "convert" to GIMP built-in sRGB), then go to `Image > Color management > Discard Color Profile`. Then you get something with believable colors. This file's embedded profile is just broken beyond measures.

> And may I ask which example JPEG2000 you used instead for testing?

This is like eons ago. I have no idea what other files I may have used. I remember I used a bunch of files, though this format doesn't have that many samples around. If I recall, the OpenJPEG project must also have some sample files in a repository.

**Comment by jubalh** [comment]
@Jehan thanks!

**Comment by ValZapod** [comment]
> looks like the description tag is missing. I am not sure if this is a mandatory tag for a well-formed profile

IT MUST be missing! Unicode data types are used to store the profile description and copyright information. These are v4 data types which are undefined in v2. The profile relax.jp2 uses v2.2.0.  http://www.color.org/security/android_bug.xalter

Here is the profile. you can play with it using ICC manager. [relax.icc.txt](https://github.com/uclouvain/openjpeg/files/5724885/relax.icc.txt)
 Looks like it is indeed corrupted in XYZ coordinates. Which means you must assume sRGB, which Gimp does not do. Instead, it, as you said, f*cks up the picture, because it somehow thinks it is not bgra (yeah, with alpha, not brg24) but yuv444p16le or something like that.
![image](https://user-images.githubusercontent.com/31514790/102787317-f1cdc980-43b1-11eb-98aa-4067c8a91c72.png)

**Comment by ValZapod** [comment]
The original file though has the following values http://www.fnordware.com/j2k/relax-orig.png 


```
Bit Depth                       : 8
Color Type                      : RGB
Compression                     : Deflate/Inflate
Filter                          : Adaptive
Interlace                       : Noninterlaced
Gamma                           : 1.8008
White Point X                   : 0.3127
White Point Y                   : 0.32899
Red X                           : 0.62502
Red Y                           : 0.34001
Green X                         : 0.28
Green Y                         : 0.595
Blue X                          : 0.155
Blue Y                          : 0.07001
Pixels Per Unit X               : 2834
Pixels Per Unit Y               : 2834
Pixel Units                     : meters
Modify Date                     : 2002:05:30 22:46:25
Saved As                        : Waterfall(16bit).png
Software                        : Adobe Photoshop
Writer                          : SuperPNG by Brendan Bolles
Image Size                      : 400x300
Megapixels                      : 0.120

```

that is, D65 with almost sRGB color primaries. Just saying.

**Comment by ValZapod** [comment]
Also, BTW gamma of 1.8 is correctly encoded as such in all 3 color tone response curves of relax.icc, which is good. Atas least gamma should be 1.8, IMHO.

**Comment by ValZapod** [comment]
> As for e-YCC, I just cannot find any reference of it anywhere

Because there is no such thing. It is actually called e-sYCC and is continuation of sYCC. It is defined in PIMA 7667 Annex B.
